src/ILCompiler.WebAssembly/src/CodeGen/ILToWebAssemblyImporter.cs

```diff
@@ -1120,31 +1131,44 @@ private void AddVirtualMethodReference(MethodDesc method)
         {
             _dependencies.Add(_compilation.NodeFactory.VirtualMethodUse(method));
         }
-
+        static Dictionary<string, MethodDesc> _pinvokeMap = new Dictionary<string, MethodDesc>();
         private void ImportRawPInvoke(MethodDesc method)
         {
-            LLVMValueRef nativeFunc = LLVM.GetNamedFunction(Module, method.Name);
-
             //emscripten dies if this is output because its expected to have i32, i32, i64. But the runtime has defined it as i8*, i8*, i64
             if (method.Name == "memmove")
                 throw new NotImplementedException();
 
+            string realMethodName = method.Name;
+            if(!method.IsPInvoke && method is TypeSystem.Ecma.EcmaMethod)
+            {
+                realMethodName = ((TypeSystem.Ecma.EcmaMethod)method).GetRuntimeImportName() ?? method.Name;
+            }
+            MethodDesc existantDesc;
+            LLVMValueRef nativeFunc;
+            LLVMValueRef realNativeFunc = LLVM.GetNamedFunction(Module, realMethodName);
+            if (_pinvokeMap.TryGetValue(realMethodName, out existantDesc))
+            {
+                if (existantDesc != method)
+                {
+                    // Set up native parameter types
+                    nativeFunc = MakeExternFunction(method, realMethodName, realNativeFunc);
```

Comment:
I'm not sure I understand how this helps the issue you described. Wouldn't we still end up building multiple extern functions if there are multiple P/Invokes?

Reply:
MakeExternFunction will cast the existing function pointer when the signatures don't match. There is a bit of a unstated dependency here on RawPInvoke != PInvoke, meaning that as long as we're only doing this for the actual underlying function pointer it should work out. If we did this trick without that distinction we would get potentially wrong marshaling behavior.